sdks/java/extensions/sbe/src/main/java/org/apache/beam/sdk/extensions/sbe/SbeSchema.java

```diff
@@ -0,0 +1,252 @@
+    private static final long serialVersionUID = 1L;
+
+    public abstract Boolean useMoreBitsForUint8();
+
+    public abstract Boolean useMoreBitsForUint16();
+
+    public abstract Boolean useMoreBitsForUint32();
+
+    public abstract Boolean useStringForUint64();
+
+    /**
+     * Returns options for using the same bit size for all unsigned types.
+     *
+     * <p>This means that if an unsigned value from SBE comes in with a value outside the signed
+     * range, then the negative equivalent (in terms of bits) will be used.
+     */
+    public static UnsignedOptions usingSameBitSize() {
+      return UnsignedOptions.builder()
+          .setUseMoreBitsForUint8(false)
+          .setUseMoreBitsForUint16(false)
+          .setUseMoreBitsForUint32(false)
+          .setUseStringForUint64(false)
+          .build();
+    }
+
+    /**
+     * Returns options for using a higher bit count for unsigned types.
+     *
+     * <p>This means that if an unsigned value is encountered, it will always use the higher bit
+     * count, even if that higher bit count is unnecessary. However, this means that if it is
+     * necessary, then the proper value will be returned rather than the negative equivalent (in
+     * terms of bits).
+     *
+     * <p>The {@code includeUint64} controls the behavior of 64-bit values, since no properly
+     * higher-bit-numeric type exists. If true, then this will be converted into a string that must
+     * be parsed if intended to be used as a number. Otherwise, it will still be a 64-bit type and
+     * may return negative values.
+     */
+    public static UnsignedOptions usingHigherBitSize(boolean includeUint64) {
+      return UnsignedOptions.builder()
+          .setUseMoreBitsForUint8(true)
+          .setUseMoreBitsForUint16(true)
+          .setUseMoreBitsForUint32(true)
+          .setUseStringForUint64(includeUint64)
+          .build();
+    }
+
+    public static Builder builder() {
+      return new AutoValue_SbeSchema_UnsignedOptions.Builder();
+    }
+
+    /** Builder for {@link UnsignedOptions}. */
+    @AutoValue.Builder
+    public abstract static class Builder {
+
+      public abstract Builder setUseMoreBitsForUint8(Boolean value);
+
+      public abstract Builder setUseMoreBitsForUint16(Boolean value);
+
+      public abstract Builder setUseMoreBitsForUint32(Boolean value);
+
+      public abstract Builder setUseStringForUint64(Boolean value);
+
+      public abstract UnsignedOptions build();
+    }
+  }
+
+  /**
+   * Options for configuring schema generation from an {@link Ir}.
+   *
+   * <p>The default options make the following assumptions:
+   *
+   * <ul>
+   *   <p>There is only message in the XML schema. In order to override this, either {@link
+   *   IrOptions#messageId()} or {@link IrOptions#messageName()} must be set, but not both.
+   * </ul>
+   */
+  @AutoValue
+  public abstract static class IrOptions implements Serializable {
```

Comment:
I'm not sure if this is your intention or not but these could be mapped to/from table params in the PayloadSerializerProvider: https://github.com/apache/beam/blob/30b440faa78f66fefd9b64be743d11021c998e82/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/payloads/PayloadSerializerProvider.java#L36

Reply:
At least for the initial option, I don't think that would work. An IR could contain many messages, but a `Schema` would be for a single one. These initial options of id and name are needed to generate the `Schema` object in case there's more than one message in the IR.

Reply:
Yes so the `IrOptions` are critical for getting a single message to map the single schema to, right?

Note that in some `getSerializer` implementations the schema is inferred, e.g. for proto we get it from the proto class: https://github.com/apache/beam/blob/30b440faa78f66fefd9b64be743d11021c998e82/sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoPayloadSerializerProvider.java#L73-L75

and the schema passed as an argument is just verified to be equal (we could and probably should make the schema allowed to be null in this case).

This is definitely out of scope for this PR though :)

Reply:
> Yes so the IrOptions are critical for getting a single message to map the single schema to, right?

Well, I thought it did, but looking over at the [OTF example](https://github.com/real-logic/simple-binary-encoding/blob/master/sbe-samples/src/main/java/uk/co/real_logic/sbe/examples/OtfExample.java), the solution is actually to wait until after decoding the message header to select the relevant message tokens. So in the deserialization PTransform, we would select the right schema dynamically.

The problem there is that it won't be compatible with the `PayloadSerializerProvider` interface, but we can probably get around that by letting a user select a particular schema themselves if they know beforehand that a particular source will only provide messages of a single type. Otherwise, it would select the schema dynamically.